clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp

```diff
@@ -354,6 +358,460 @@ fatbinary(ArrayRef<std::pair<StringRef, StringRef>> InputFiles,
+  }
+  return *OutputFileOrErr;
+}
+
+// This routine is used to unbundle all device library files that will be
+// linked with input device codes.
+static Error
+unbundleSYCLDeviceLibs(const SmallVector<std::string, 16> &Files,
+                       SmallVector<std::string, 16> &UnbundledFiles,
+                       const ArgList &Args) {
+  for (auto &Filename : Files) {
+    if (!sys::fs::exists(Filename) || sys::fs::is_directory(Filename))
+      continue;
+    // Run unbundler
+    auto UnbundledFile = sycl::unbundle(Filename, Args);
+    if (!UnbundledFile)
+      return UnbundledFile.takeError();
+    UnbundledFiles.push_back((*UnbundledFile).str());
+  }
+  return Error::success();
+}
+
+// Link all SYCL input files into one before adding device library files.
+Expected<StringRef> linkDeviceInputFiles(SmallVectorImpl<StringRef> &InputFiles,
+                                         const ArgList &Args) {
+  llvm::TimeTraceScope TimeScope("LinkDeviceInputFiles");
+
+  Expected<std::string> LLVMLinkPath =
+      findProgram("llvm-link", {getMainExecutable("llvm-link")});
+  if (!LLVMLinkPath)
+    return LLVMLinkPath.takeError();
+
+  // Create a new file to write the linked device file to.
+  auto OutFileOrErr =
+      createOutputFile(sys::path::filename(ExecutableName), "bc");
+  if (!OutFileOrErr)
+    return OutFileOrErr.takeError();
+
+  SmallVector<StringRef, 8> CmdArgs;
+  CmdArgs.push_back(*LLVMLinkPath);
+  for (auto File : InputFiles)
+    CmdArgs.push_back(File);
+  CmdArgs.push_back("-o");
+  CmdArgs.push_back(*OutFileOrErr);
+  CmdArgs.push_back("--suppress-warnings");
+  if (Error Err = executeCommands(*LLVMLinkPath, CmdArgs))
+    return std::move(Err);
+  return *OutFileOrErr;
+}
+
+// Link all device library files and input file into one.
+static Expected<StringRef>
+linkDeviceLibFiles(SmallVectorImpl<StringRef> &InputFiles,
+                   const ArgList &Args) {
+  llvm::TimeTraceScope TimeScope("LinkDeviceLibraryFiles");
+
+  Expected<std::string> LLVMLinkPath =
+      findProgram("llvm-link", {getMainExecutable("llvm-link")});
+  if (!LLVMLinkPath)
+    return LLVMLinkPath.takeError();
+
+  // Create a new file to write the linked device file to.
+  auto OutFileOrErr =
+      createOutputFile(sys::path::filename(ExecutableName), "bc");
+  if (!OutFileOrErr)
+    return OutFileOrErr.takeError();
+
+  SmallVector<StringRef, 8> CmdArgs;
+  CmdArgs.push_back(*LLVMLinkPath);
+  CmdArgs.push_back("-only-needed");
+  for (auto &File : InputFiles)
+    CmdArgs.push_back(File);
+  CmdArgs.push_back("-o");
+  CmdArgs.push_back(*OutFileOrErr);
+  CmdArgs.push_back("--suppress-warnings");
+  if (Error Err = executeCommands(*LLVMLinkPath, CmdArgs))
+    return std::move(Err);
+  return *OutFileOrErr;
+}
+
+static Expected<StringRef> linkDevice(ArrayRef<StringRef> InputFiles,
+                                      const ArgList &Args) {
+  SmallVector<StringRef, 16> InputFilesVec;
+  for (StringRef InputFile : InputFiles)
```

Comment:
Nit, I think we can do a iterator insert here instead of a loop

```
InputFilesVec.insert(InputFilesVec.begin(), InputFiles.begin(), InputFiles.end());
```

Reply:
I'm not sure emplace_back is doing anything for us here. Usually with emplace_back you give it the constructor arguments of the class to be constructed, but here the vector value type and argument type are equal, so I think it's the same as push_back